core/survey/nodeDef.js

```diff
@@ -256,6 +256,33 @@ export const getChainIndex = getPropOrDraftAdvanced(keysPropsAdvanced.index, 0)
 export const getActive = getPropOrDraftAdvanced(keysPropsAdvanced.active, false)
 export const getScript = getPropOrDraftAdvanced(keysPropsAdvanced.script, '')
 
+// Analysis
+
+export const getNodeDefsWithColumnNames = (nodeDefs) => {
+  return nodeDefs.flatMap((nodeDef) => {
+    const name = getName(nodeDef)
+    if (isCode(nodeDef)) {
+      return [
+        {
+          columnName: `${name}_code`,
+          nodeDef,
+        },
+        { columnName: `${name}_label`, nodeDef },
+      ]
+    }
+    return { columnName: name, nodeDef }
+  })
+}
+
+export const getNodeDefsColumnNames = (nodeDefs) =>
+  getNodeDefsWithColumnNames(nodeDefs).map(({ columnName }) => columnName)
+
+export const getNodeDefsByColumnNames = (nodeDefs) =>
+  getNodeDefsWithColumnNames(nodeDefs).reduce(
+    (_nodeDefs, { columnName, nodeDef }) => ({ ..._nodeDefs, [columnName]: nodeDef }),
+    {}
+  )
+
```

Comment:
there are already methods in surveyRdb modules to get column names by node def

Reply:
`
 get columnNodeDefs() {
    return R.pipe(
      R.ifElse(
        NodeDef.isEntity,
        R.pipe(
          () => Survey.getNodeDefChildren(this.nodeDef, NodeDef.isAnalysis(this.nodeDef))(this.survey),
          R.filter(NodeDef.isSingleAttribute),
          R.sortBy(R.ascend(R.prop('id')))
        ),
        (nodeDef) => [nodeDef] // Multiple attr table
      ),
      R.map((nodeDefColumn) => new ColumnNodeDef(this, nodeDefColumn))
    )(this.nodeDef)
  }
  
  `
  
  ###############

  `
 import * as R from 'ramda'
import camelize from 'camelize'
import * as toSnakeCase from 'to-snake-case'

import * as NodeDef from '../../../../../core/survey/nodeDef'
import * as SQL from '../../sql'

const { nodeDefType } = NodeDef

const colNamesSuffixByType = {
  [nodeDefType.code]: ['code', 'label'],
  [nodeDefType.taxon]: ['code', 'scientific_name'],
  [nodeDefType.file]: ['file_uuid', 'file_name'],
}

const colTypesByType = {
  [nodeDefType.boolean]: [SQL.types.varchar],
  [nodeDefType.code]: [SQL.types.varchar, SQL.types.varchar],
  [nodeDefType.coordinate]: [SQL.types.geometryPoint],
  [nodeDefType.date]: [SQL.types.date],
  [nodeDefType.decimal]: [SQL.types.decimal],
  [nodeDefType.entity]: [SQL.types.uuid],
  [nodeDefType.file]: [SQL.types.uuid, SQL.types.varchar],
  [nodeDefType.integer]: [SQL.types.bigint],
  [nodeDefType.taxon]: [SQL.types.varchar, SQL.types.varchar],
  [nodeDefType.text]: [SQL.types.varchar],
  [nodeDefType.time]: [SQL.types.time],
}

const getColNames = (nodeDef) => {
  const nodeDefName = NodeDef.getName(nodeDef)
  const colsSuffix = colNamesSuffixByType[NodeDef.getType(nodeDef)]
  if (colsSuffix) {
    return colsSuffix.map((col) => `${nodeDefName}_${col}`)
  }
  if (NodeDef.isEntity(nodeDef)) {
    return [`${nodeDefName}_uuid`]
  }
  return [nodeDefName]
}

const extractColName = ({ nodeDef, colName }) =>
  camelize(colName.replace(`${toSnakeCase(NodeDef.getName(nodeDef))}_`, ''))

/**
 * A nodeDef data table column.
 *
 * @typedef {object} module:arena.ColumnNodeDef
 */
export default class ColumnNodeDef {
  constructor(table, nodeDef) {
    this._table = table
    this._nodeDef = nodeDef
    this._names = getColNames(nodeDef)
    this._types = colTypesByType[NodeDef.getType(nodeDef)]
  }

  get table() {
    return this._table
  }

  get nodeDef() {
    return this._nodeDef
  }

  get names() {
    return this._names
  }

  get name() {
    return this.names[0]
  }

  get namesFull() {
    return SQL.addAlias(this.table.alias, this.names)
  }

  get nameFull() {
    return this.namesFull[0]
  }

  get types() {
    return this._types
  }
}

ColumnNodeDef.getColNames = getColNames
ColumnNodeDef.getColName = R.pipe(ColumnNodeDef.getColNames, R.head)
ColumnNodeDef.extractColName = extractColName

`

Reply:
maybe we have to extract this to use outside of classes